README
NAME
    Perl::Tidy::Sweetened - Tweaks to Perl::Tidy to support some syntactic
    sugar

VERSION
    version 0.18

DESCRIPTION
    There are a number of modules on CPAN that allow users to write their
    classes with a more "modern" syntax. These tools eliminate the need to
    shift off $self, can support type checking and offer other improvements.
    Unfortunately, they can break the support tools that the Perl community
    has come to rely on. This module attempts to work around those issues.

    The module uses Perl::Tidy's "prefilter" and "postfilter" hooks to
    support "method" and "func" keywords, including the (possibly
    multi-line) parameter lists. This is quite an ugly hack, but it is the
    recommended method of supporting these new keywords (see the 2010-12-17
    entry in the Perl::Tidy CHANGES
    <https://metacpan.org/source/SHANCOCK/Perl-Tidy-20120714/CHANGES> file).
    The resulting formatted code will leave the parameter lists untouched.

    "Perl::Tidy::Sweetened" attempts to support the syntax outlined in:

    *   Method::Signature::Simple

    *   MooseX::Method::Signatures

    *   MooseX::Declare

SEE ALSO
    Perl::Tidy

    The idea and much of original code taken from Jonathan Swartz' blog
    <http://www.openswartz.com/2010/12/19/perltidy-and-method-happy-together
    />.

BUGS
    Please report any bugs or suggestions at
    <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Perl-Tidy-Sweetened>

AUTHOR
    Mark Grimes, <mgrimes@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Mark Grimes, <mgrimes@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changes for Perl::Tidy::Sweetend

1.19 - Mon Feb 20 21:26:40 EST 2023
    Updates Perl::Tidy dependency (GH#23)
    Thanks to @shancock9 this release of Perl::Tidy resolves @dboehmer's
    issues where 'Use of uninitialized value $body' error messages are
    printed.

    Updates the repo and bugtracker urls to use https (PR #22)
    Thanks @szabgab for the PR.
    [Upgrade recommended.]

1.18 - Sun Jan 31 12:24:35 EST 2021
    Fixes tests to handle more changes in spacing in Perl::Tidy versions
    [Only changes tests. No need to upgrade.]

1.17 - Sun Jan 24 11:36:21 EST 2021
    Handles errors in closing sub comments when there are 10+ subs.
    Fixes GH#20. Thanks to @fpostma72 for the bug report and PR.
    
    Perl-Tidy-20200907 changed an indentation of side comments. This patch
    adjusts the tests.
    Fixes GH#18. Thanks to @ppisar for the bug report and PR.
    
    Respects the -cscp argument.
    [Upgrade recommended if you use closing sub comments.]

1.16 - Wed Apr  8 11:57:37 EDT 2020
    Adds support for 'around/before/after' keywords (GH#17)
    Thanks to witoldwegner for the PR.
    [Upgrade recommended if you use these method modifiers.]

1.15 - Tue Aug 28 09:15:20 EDT 2018
    Adds basic support for kavorka traits (is/does/but) (GH#16)
    Update the perl versions to test with travis (+5.28, +5.26, -5.21)
    Thanks to Peter Roberts (@pwr22) for the PR. (GH#16).
    [Upgrade recommended if you use Kavorka.]

1.14 - Wed Mar 28 17:36:19 EDT 2018
    Adds support for Kavorka's basic keywords.
    Thanks to Peter Roberts (@pwr22) for the PR. (GH#13).
    [Upgrade recommended if you use Kavorka.]

1.13 - Thu May 18 12:47:46 EDT 2017
    Adds support for closing side comments (GH#12)
    Thanks to Aaron Priven for the bug report
    [Upgrade recommended if you use the -csc flag.]

1.12 - Tue Oct  4 15:56:53 EDT 2016
    Add support to extend classes with :: (e.g, Object::Person).
    Thanks to @pblaberge for the PR.
    [Upgrade recommended.]

1.11 - Fri Apr 15 09:29:03 EDT 2016
    Sends Perl::Tidy errors during testing to a string instead of fs (GH#9)
    - By default Perl::Tidy writes a perltidy.ERR file when it encounters
      errors. Perl::Tidy::Sweetened tests remove the error file then check
      to see if it is created after each test. This leads to a race
      condition with running in parallel (ie, prove -j 9). This commit makes
      Perl::Tidy write errors to a string buffer instead. Thanks @shlomif
      for the bug report (GH#9).
    [Only changes tests. No need to upgrade.]

1.10 - Sun Mar 13 13:14:01 EDT 2016
    Skips test that fails with Perl::Tidy 20160301 due to a Perl::Tidy bug
    - Adds test-all script to run against multiple versions of Perl::Tidy
    - Moves display of Perl::Tidy version to t/00-version.t
    Fixes typo in cpan badge URL
    [Only changes tests. No need to upgrade.]

1.09 - Fri Mar 11 21:27:42 EST 2016
    Converts from Module::Build to Module::Built::Tiny and the mbtiny tool
    Adds perltidy-sweet script to distribution
    Fixes issues with badges
    [Optional upgrade]

1.08 - Fri Mar 11 20:55:29 EST 2016
    Updates tests to support Perl::Tidy 20160301 and later
    - New version of Perl::Tidy has some support for new syntax. This is
      great, but it has some minor impact on spacing in our tests.
    [Upgrade recommended if you are using newer versions of Perl::Tidy]

1.07 - Wed Jan 27 09:36:42 EST 2016
    Adds new podweaver sections
    - Bugs in pod now point to GH issues
    - Adds contributors list to pod
    - Adds link to github in pod
    [No need to upgrade]

1.06 - Tue Jan 26 17:21:34 EST 2016
    Ignore 'method' functions (not pseudo-keywords) like those in
      MooseX::Role::Parameterized (PR#6, RT#106464).
    Use faster container-based infrastructure for Travis (PR#7).
    Both of these patch were contributed by @oschwald. Thanks!
    [Upgrade recommended if you use MooseX::Role::Parameterized]

1.05 - Sat Aug 22 10:07:24 EDT 2015
    Changes placeholder from comment at end of line, to munged sub/pkg name.
    So, instead of turning:

        method test_method_a ($self) {

    into

        sub test_method_a { # SUB1

    we now turn it into

        sub SUB1_method_a {

    This eliminates issues with Perl::Tidy moving that trailing comment, and
    attempts to preserve the length of the method name so any changes to
    formatting due to name length will be consistent.

    Bug fixes and improvements to tests:
    - Bug fix: RT#106398 "long single line subs" should be handled correctly.
    - Adds (failing) test for RT#106464
    - Adds checks for creation of perltidy.ERR file as part of tests
    - Adds simpler means of specifying TODO tests

    TODO:
    - Would like to preserve the *length* of the clauses in order to allow
      Perl::Tidy to format them. This could probably be done with prototypes
      for sub, but not sure what you do with packages.

    [This upgrade is recommended if you are experiencing issues with
    "# SUB_" comments being left in your code.]

1.04 - Fri Mar 20 09:16:11 EDT 2015
    This release fixes bug reported in RT#102815.
    - We had been overwriting the result of the -sbl and -bl flags, which
      control line breaks before opening subroutine brace.  Thanks to Serge
      Colle for the bug report.
    - This fix makes changes to the way the post_filter parses a line. Please
      test this with your codebase and report any bugs.
    [This upgrade is recommended if you use the -sbl or -bl flags.]

1.03 - Sat Dec 20 16:53:27 EST 2014
    Adds the perltidy-sweet command
    - More clearly indicates how this differs from perltidy than perltidier.
      Thanks to Dave Rolsky (autarch) for the suggestion.
    - Keeps perltidier command as many users are already used to this.

1.02 - Tue Dec  9 12:28:49 EST 2014
    Fixes handling of 5.20 experimental signatures
    - It looked like Perl::Tidy would handle the new 5.20 experimental

t/lib/TidierTests.pm
package TidierTests;

use strict;
use warnings;
use Perl::Tidy::Sweetened;
use Test::Most;

use Exporter;
@TidierTests::ISA    = qw(Exporter);
@TidierTests::EXPORT = qw(run_test $indent);

sub run_test {
    my ( $raw, $expected, $msg, $todo, @args ) = @_;

    my $errorbuffer = '';
    my @tidied;
    ## warn "# -nsyn -ce -npro -l=60 " . join( ' ', @args ), "\n";
    Perl::Tidy::Sweetened::perltidy(
        source      => \$raw,
        destination => \@tidied,
        perltidyrc  => undef,
        errorfile   => \$errorbuffer,
        argv        => '-nsyn -ce -npro -l=60 ' . join( ' ', @args ),
    );
    my $tidied = join '', @tidied;

    if ($todo) {

      TODO: {
            # Works with Test::More prior to 1.301001_?
            local $TidierTests::TODO = 'Not implmented';

            # Works with Test::More after 1.301001_021
            local $TODO = 'Not implmented';

            return check_test( $tidied, $expected, $errorbuffer, $msg );
        }

    } else {
        return check_test( $tidied, $expected, $errorbuffer, $msg );
    }
}

sub check_test {
    my ( $tidied, $expected, $errors, $msg ) = @_;

    my $ok_log = is( $errors, '', "$msg: no errors" );
    my $ok_tidy = eq_or_diff( $tidied, $expected, "$msg: matches" );

    return $ok_log && $ok_tidy;
}

require Perl::Tidy;
our $indent = ' ' x (($Perl::Tidy::VERSION < 20200907) ? 6 : 4);

1;


lib/Perl/Tidy/Sweet.pm
package Perl::Tidy::Sweet;

# ABSTRACT: Tweaks to Perl::Tidy to support some syntactic sugar

use strict;
use warnings;
use base 'Perl::Tidy::Sweetened';

our $VERSION = '1.19';

1;

__END__

=pod

=head1 NAME

Perl::Tidy::Sweet - Tweaks to Perl::Tidy to support some syntactic sugar

=head1 VERSION

version 1.19

=head1 STATUS

=for html <a href="https://travis-ci.org/mvgrimes/Perl-Tidy-Sweetened"><img src="https://travis-ci.org/mvgrimes/Perl-Tidy-Sweetened.svg?branch=master" alt="Build Status"></a>
<a href="https://metacpan.org/pod/Perl::Tidy::Sweetened"><img alt="CPAN version" src="https://badge.fury.io/pl/Perl-Tidy-Sweetened.svg" /></a>

=head1 DESCRIPTION

There are a number of modules on CPAN that allow users to write their classes
with a more "modern" syntax. These tools eliminate the need to shift off
C<$self>, can support type checking and offer other improvements.
Unfortunately, they can break the support tools that the Perl community has
come to rely on. This module attempts to work around those issues.

The module uses
L<Perl::Tidy>'s C<prefilter> and C<postfilter> hooks to support C<method> and
C<func> keywords, including the (possibly multi-line) parameter lists. This is
quite an ugly hack, but it is the recommended method of supporting these new
keywords (see the 2010-12-17 entry in the Perl::Tidy
L<CHANGES|https://metacpan.org/source/SHANCOCK/Perl-Tidy-20120714/CHANGES>
file). B<The resulting formatted code will leave the parameter lists untouched.>

C<Perl::Tidy::Sweetened> attempts to support the syntax outlined in the
following modules, but most of the new syntax styles should work:

=over

=item * p5-mop

=item * Method::Signatures::Simple

=item * MooseX::Method::Signatures

=item * MooseX::Declare

=item * Moops

=item * perl 5.20 signatures

=item * Kavorka

=back

=head1 SEE ALSO

L<Perl::Tidy>

=head1 AUTHOR

Mark Grimes <mgrimes@cpan.org>

=head1 SOURCE

Source repository is at L<https://github.com/mvgrimes/Perl-Tidy-Sweetened>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://github.com/mvgrimes/Perl-Tidy-Sweetened/issues>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2023 by Mark Grimes <mgrimes@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Perl/Tidy/Sweetend.pm
package Perl::Tidy::Sweetend;

use strict;
use warnings;
use Perl::Tidy qw();

our $VERSION = '0.11';

sub perltidy {
    return Perl::Tidy::perltidy(
        prefilter  => \&prefilter,
        postfilter => \&postfilter,
        @_
    );
}

sub prefilter {
    $_ = $_[0];

    no warnings 'uninitialized';

    # Convert func xxxx (signature) -> sub
    s/^\s*\Kfunc\s+(\w+)\s+(\([^)]*\))?(.*?)$/sub $1 $3 \#__FUNC $2/gm;

    # Convert method -> sub
    s/^\s*\Kmethod\s+(\w+)\s*(\([^)]*\))?(.*?)$/sub $1 $3 \#__METHOD $2/gm;

    # warn ">>>>>>>>>\n$_";
    return $_;
}

sub postfilter {
    $_ = $_[0];

    no warnings 'uninitialized';
    # warn "========\n$_";

    # Convert func back
    s/^\s*\Ksub\s+(\w+)\b(.*?) *\#__FUNC( \([^)]+\))? */func $1$3$2/gm;

    # Convert sub and signature back to method
    s/^\s*\Ksub\s+(\w+)\b(.*?) *\#__METHOD( \([^)]+\))? */method $1$3$2/gm;

    # Convert sub back to method
    # s/^sub (.*?)\s* \#__METHOD/method $1/gm;

    # Remove spaces from method signature
    # s/^method (\w+) \(([^)]+?):\s/method $1 ($2:/gm;

    return $_;
}

1;

__END__

=pod

=head1 NAME

Perl::Tidy::Sweetend - Tweaks to Perl::Tidy to work with modern method systanx

=head1 DESCRIPTION

Fixes...

=head1 SEE ALSO

L<Perl::Tidy>

=head1 BUGS

Please report any bugs or suggestions at 
L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Perl-Tidy-Sweetend>

=head1 AUTHOR

Mark Grimes, E<lt>mgrimes@cpan.orgE<gt>

Ideal and much of original code taken from Jonathan Swartz blog
(http://www.openswartz.com/2010/12/19/perltidy-and-method-happy-together/).

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 by Mark Grimes

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.2 or,
at your option, any later version of Perl 5 you may have available.



lib/Perl/Tidy/Sweetened/Pluggable.pm
package Perl::Tidy::Sweetened::Pluggable;

# ABSTRACT: Simple object to facilitate a pluggable filter architecture

use strict;
use warnings;

our $VERSION = '1.19';

sub new {
    my ( $class, %args ) = @_;
    return bless {%args}, $class;
}

sub filters { return ( $_[0]->{filters} ||= [] ) }

sub args { return ( $_[0]->{args} ||= {} ) }

sub add_filter {
    my $self = shift;
    push @{ $self->filters }, @_;
}

sub prefilter {
    my ( $self, $code ) = @_;
    for my $filter ( @{ $self->filters } ) {
        $code = $filter->prefilter($code, $self->args);
    }

    # warn "After prefilter, before tidy\n";
    # warn $code;

    return $code;
}

sub postfilter {
    my ( $self, $code ) = @_;

    # warn "After tidy, before postfilter\n";
    # warn $code;

    for my $filter ( @{ $self->filters } ) {
        $code = $filter->postfilter($code, $self->args);
    }
    return $code;
}

sub add_args {
    my ( $self,       $args ) = @_;
    my ( $body_parts, $msg )  = Perl::Tidy::parse_args($args);
    $self->{args} =
      { map { my ( $k, $v ) = split /=/, $_, 2; ( $k => $v ) } @$body_parts };
}

1;

__END__

=pod

=head1 NAME

Perl::Tidy::Sweetened::Pluggable - Simple object to facilitate a pluggable filter architecture

=head1 VERSION

version 1.19

=head1 SYNOPSIS

    our $plugins = Perl::Tidy::Sweetened::Pluggable->new();

    $plugins->add_filter(
        Perl::Tidy::Sweetened::Keyword::SubSignature->new(
            keyword => 'method',
            marker  => 'METHOD',
        ) );

    $plugins->prefilter( $code );
    $plugins->postfilter( $code );

=head1 DESCRIPTION

Builds a pluggable, chainable list of filters.

=head1 AUTHOR

Mark Grimes <mgrimes@cpan.org>

=head1 SOURCE

Source repository is at L<https://github.com/mvgrimes/Perl-Tidy-Sweetened>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://github.com/mvgrimes/Perl-Tidy-Sweetened/issues>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2023 by Mark Grimes <mgrimes@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Perl/Tidy/Sweetened/Keyword/Block.pm
package Perl::Tidy::Sweetened::Keyword::Block;

# ABSTRACT: Perl::Tidy::Sweetened filter plugin to define new subroutine and class keywords

use 5.010;    # Needed for balanced parens matching with qr/(?-1)/
use strict;
use warnings;
use Carp;
$|++;

our $VERSION = '1.19';

# Regex to match balanced params. Reproduced from Regexp::Common to avoid
# adding a non-core dependency.
#   $RE{balanced}{-parens=>'()'};
# The (?-1) construct requires 5.010
our $Paren = '(?:((?:\((?:(?>[^\(\)]+)|(?-1))*\))))';

sub new {
    my ( $class, %args ) = @_;
    croak 'keyword not specified'     if not exists $args{keyword};
    croak 'marker not specified'      if not exists $args{marker};
    croak 'replacement not specified' if not exists $args{replacement};
    $args{clauses} = [] unless exists $args{clauses};
    return bless {%args}, $class;
}

sub keyword     { return $_[0]->{keyword} }
sub marker      { return $_[0]->{marker} }
sub replacement { return $_[0]->{replacement} }

sub emit_placeholder {
    my ( $self, $subname, $brace, $clauses ) = @_;

    # Store the signature and returns() for later use
    my $id = sprintf "%03d", $self->{counter}++;
    $self->{store_clause}->{$id} = $clauses;
    $self->{store_sub}->{$id}    = $subname;

    # Turns 'my_method_name' into 'SUB004hod_name'
    my $marker = $self->marker . $id;
    substr( $subname, 0, length($marker), $marker );

    return sprintf '%s %s %s', $self->replacement, $marker, $brace;
}

sub emit_keyword {
    my ( $self, $brace, $id ) = @_;

    # Get the signature and returns() from store
    my $clauses = $self->{store_clause}->{$id};

    my $subname = $self->{store_sub}->{$id};

    # Combine clauses (parameter list, returns(), etc) into a string separate
    # each with a space and lead with a space if there are any
    my $clause = join ' ', grep { length $_ } @$clauses;

    # FIXME: This forces space between sub/func/method name and clauses list
    # (ignores perltidy settings)
    $clause = ' ' . $clause if length $clause;

    return sprintf '%s %s%s%s', $self->keyword, $subname, $clause, $brace;
}

sub emit_csc {
    my ( $self, $id, $cscp ) = @_;
    return sprintf "%s %s %s", $cscp, $self->keyword, $self->{store_sub}->{$_};
}

sub clauses {
    my $self = shift;

    # Create a regex (as a string) for all the clauses (ie, parameter list,
    # returns(), etc).
    my $clause_re = '';
    my $i         = 0;
    for my $clause ( @{ $self->{clauses} } ) {
        $clause =~ s{PAREN}{$Paren}g;

        $clause_re .= "(?<clause_$i>  $clause ) \\s* \n";
        $i++;
    }

    return $clause_re;
}

sub identifier {    # method or package identifier
    my $self = shift;

    return '\w+ (?: ::\w+ )*';    # words, possibly separated by ::
}

sub prefilter {
    my ( $self, $code ) = @_;
    my $keyword = $self->keyword;
    my $subname = $self->identifier;

    $code =~ s{
        ^\s*\K                    # okay to have leading whitespace (preserve)
        $keyword             \s+  # the "func/method" keyword
        (?<subname> $subname)     # the function name or class name (needs ::)
        (?!\w|\s*=>) \s*          # check to make sure this isn't a sub call with params
        @{[ $self->clauses ]}     # any clauses defined (ie, a parameter list)
        (?<brace> .*?)            # anything else (ie, comments) including brace
        $
    }{
        my $i = 0;
        my $clauses = [];
        while( exists $+{"clause_$i"} ){
            ## warn "# clause_$i: " . $+{"clause_$i"} . "\n";
            push @$clauses, $+{"clause_$i"};
            $i++;
        }
        $self->emit_placeholder( $+{subname}, $+{brace}, $clauses )
    }egmx;

    return $code;
}

sub postfilter {
    my ( $self, $code, $args ) = @_;
    my $marker      = $self->marker;
    my $replacement = $self->replacement;
    my $subname     = $self->identifier;
    my @ids;

    # Convert back to method
    $code =~ s{
        ^\s*\K                     # preserve leading whitespace
        $replacement          \s+  # keyword was converted to sub/package
        $marker                    #
        (?<id> \d\d\d)             # the identifier
        [\w:]* \b                  # the rest of the orignal sub/package name
        (?<newline> \n? \s* )      # possible newline and indentation
        (?<brace>   .*?     ) [ ]* # opening brace on followed orig comments
        [ ]*                       # trailing spaces (not all whitespace)
    }{
        push @ids, $+{id};
        $self->emit_keyword( $+{newline} . $+{brace}, $+{id} );
    }egmx;

    # Restore the orig sub name when inserted via the -csc flag
    my $cscp = $args->{'-cscp'} || '## end';
    $code =~ s{
        \Q${cscp}\E \s sub \s ${marker} $_
    }{
        $self->emit_csc( $_, $cscp );
    }egx for @ids;

    return $code;
}

1;

__END__

=pod

=head1 NAME

Perl::Tidy::Sweetened::Keyword::Block - Perl::Tidy::Sweetened filter plugin to define new subroutine and class keywords

=head1 VERSION

version 1.19

=head1 SYNOPSIS

    our $plugins = Perl::Tidy::Sweetened::Pluggable->new();

    $plugins->add_filter(
        Perl::Tidy::Sweetened::Keyword::Block->new(
            keyword     => 'method',
            marker      => 'METHOD',
            replacement => 'sub',
            clauses     => [ 'PAREN?', '(returns \s* PAREN)?' ],
        ) );

=head1 DESCRIPTION

This is a Perl::Tidy::Sweetened filter which enables the definition of
arbitrary keywords for subroutines with any number of potential signature
definitions. New accepts:

=over 4

=item keyword

    keyword => 'method'

Declares a new keyword (in this example the "method" keyword).

=item marker

    marker => 'METHOD'

Provides a text marker to be used to flag the new keywords during
C<prefilter>. The source code will be filtered prior to formatting by
Perl::Tidy such that:

    method foo {
    }

is turned into:

    sub foo { # __METHOD 1
    }

=item replacement

    replacement => 'sub'

Will convert the keyword to a C<sub> as shown above.

=item clauses

    clauses => [ 'PAREN?' ]

Provides a list of strings which will be turned into a regex to capture
additional clauses. The regex will include the 'xm' flags (so be sure to escape
spaces).  The clause can be marked optional with '?'. The special text "PAREN"
can be used to capture a balanced parenthetical.

This example will capture a parameter list enclosed by parenthesis, ie:

    method foo (Int $i) {
    }

No formatting is done on the clauses at this time. The order of declaration
is significant.

=back

=head1 AUTHOR

Mark Grimes <mgrimes@cpan.org>

=head1 SOURCE

Source repository is at L<https://github.com/mvgrimes/Perl-Tidy-Sweetened>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://github.com/mvgrimes/Perl-Tidy-Sweetened/issues>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2023 by Mark Grimes <mgrimes@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Perl/Tidy/Sweetened/Variable/Twigils.pm
package Perl::Tidy::Sweetened::Variable::Twigils;

# ABSTRACT: Perl::Tidy::Sweetened filter plugin to handle twigls

use 5.010;    # Needed for balanced parens matching with qr/(?-1)/
use strict;
use warnings;
use Carp;
$|++;

our $VERSION = '1.19';

sub new {
    my ( $class, %args ) = @_;
    croak 'twigil not specified' if not exists $args{twigil};
    croak 'marker not specified' if not exists $args{marker};
    $args{clauses} = [] unless exists $args{clauses};
    return bless {%args}, $class;
}

sub twigil { return $_[0]->{twigil} }
sub marker { return $_[0]->{marker} }

sub emit_placeholder {
    my ( $self, $varname ) = @_;

    # Store the signature and returns() for later use
    my $id = $self->{counter}++;
    $self->{store}->{$id} = $varname;

    return sprintf '$__%s_%s', $self->marker, $id;
}

sub emit_twigil {
    my ( $self, $id ) = @_;

    # Get the signature and returns() from store
    my $varname = $self->{store}->{$id};

    return sprintf '%s%s', $self->twigil, $varname;
}

sub prefilter {
    my ( $self, $code ) = @_;
    my $twigil = '\\' . $self->twigil;

    $code =~ s{
        (?: ^|\s)\K               # needs to be sperated by a space
        $twigil                   # the twigil (ie, $!)
        (?<varname> \w+)          # the variable name
    }{
        $self->emit_placeholder( $+{varname} )
    }egmx;

    return $code;
}

sub postfilter {
    my ( $self, $code ) = @_;
    my $marker = $self->marker;

    # Convert back to method
    $code =~ s{
        (?: ^|\s)\K            # needs to be sperated by a space
        \$ __ $marker          # keyword was convert to package
        _ (?<id> \d+ ) \b      # the method name and a word break
    }{
        $self->emit_twigil( $+{id} );
    }egmx;

    # Check to see if tidy turned it into "sub name\n{ #..."
    $code =~ s{
        ^\s*\K                   # preserve leading whitespace
        package             \s+  # method was converted to sub
        (?<subname> \w+)\n  \s*  # the method name and a newline
        (?<brace> \{ .*?)   [ ]* # opening brace on newline followed orig comments
        \#__$marker         \s+  # our magic token
        (?<id> \d+)              # our sub identifier
        [ ]*                     # trailing spaces (not all whitespace)
    }{
        $self->emit_keyword( $+{subname}, $+{brace}, $+{id} );
    }egmx;

    return $code;
}

1;

__END__

=pod

=head1 NAME

Perl::Tidy::Sweetened::Variable::Twigils - Perl::Tidy::Sweetened filter plugin to handle twigls

=head1 VERSION

version 1.19

=head1 SYNOPSIS

    our $plugins = Perl::Tidy::Sweetened::Pluggable->new();

    $plugins->add_filter(
        Perl::Tidy::Sweetened::Variable::Twigils->new(
            twigil => '$!',
            marker => 'TWG_BANG',
        ) );

=head1 DESCRIPTION

This is a Perl::Tidy::Sweetened filter which enables the use of twigils as
defined by the L<Twigils> module.  New accepts:

=over 4

=item twigil

    twigil => '$!'

Declares a new twigil. In this case to be used as C<$!variable>.

=item marker

    marker => 'TWG_BANG'

Provides a text marker to be used to flag the new keywords during
C<prefilter>. The source code will be filtered prior to formatting by
Perl::Tidy such that:

    $!class_attribute

is turned into:

    $__TWG_BANK_1

Then back into the original twigiled variable in the C<postfilter>.

=back

=head1 AUTHOR

Mark Grimes <mgrimes@cpan.org>

=head1 SOURCE

Source repository is at L<https://github.com/mvgrimes/Perl-Tidy-Sweetened>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://github.com/mvgrimes/Perl-Tidy-Sweetened/issues>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2023 by Mark Grimes <mgrimes@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Perl/Tidy/Sweetened.pm
package Perl::Tidy::Sweetened;

# ABSTRACT: Tweaks to Perl::Tidy to support some syntactic sugar

use 5.010;
use strict;
use warnings;
use Perl::Tidy qw();

our $VERSION = '1.19';

use Perl::Tidy::Sweetened::Pluggable;
use Perl::Tidy::Sweetened::Keyword::Block;
use Perl::Tidy::Sweetened::Variable::Twigils;

our $plugins = Perl::Tidy::Sweetened::Pluggable->new();

# Create a subroutine filter for:
#    sub foo ($i, $j) {}
$plugins->add_filter(
    Perl::Tidy::Sweetened::Keyword::Block->new(
        keyword     => 'sub',
        marker      => 'SUB',
        replacement => 'sub',
        clauses     => ['PAREN'],
    ) );

# Create a subroutine filter for:
#    func foo (Int $i) returns (Bool) {}
# where both the parameter list and the returns type are optional
$plugins->add_filter(
    Perl::Tidy::Sweetened::Keyword::Block->new(
        keyword     => 'func',
        marker      => 'FUNC',
        replacement => 'sub',
        clauses     => [ 'PAREN?', '(returns \s* PAREN)?' ],
    ) );

# Create a subroutine filter for:
#    fun foo (Int $i) {}
# where the parameter list is optional
$plugins->add_filter(
    Perl::Tidy::Sweetened::Keyword::Block->new(
        keyword     => 'fun',
        marker      => 'FUN',
        replacement => 'sub',
        clauses     => ['PAREN?'],
    ) );

# Create a subroutine filter for:
#    method foo (Int $i) returns (Bool) {}
# where both the parameter list and the returns type are optional
$plugins->add_filter(
    Perl::Tidy::Sweetened::Keyword::Block->new(
        keyword     => 'method',
        marker      => 'METHOD',
        replacement => 'sub',
        clauses =>
          [ 'PAREN?', '(returns \s* PAREN)?', '(\b(?:is|but|does) \s+ \w+)?' ],
    ) );

# Create a subroutine filter for:
#    around foo (Int $i) returns (Bool) {}
#    before foo (Int $i) returns (Bool) {}
#    after foo (Int $i) returns (Bool) {}
# where both the parameter list and the returns type are optional
$plugins->add_filter(
    Perl::Tidy::Sweetened::Keyword::Block->new(
        keyword     => $_,
        marker      => uc($_),
        replacement => 'sub',
        clauses     => ['PAREN?'],
    ) ) for qw(around before after);

# Create a subroutine filter for:
#    classmethod foo (Int $i) {}
# where the parameter list is optional
$plugins->add_filter(
    Perl::Tidy::Sweetened::Keyword::Block->new(
        keyword     => 'classmethod',
        marker      => 'CLASSMETHOD',
        replacement => 'sub',
        clauses     => ['PAREN?'],
    ) );

# Create a subroutine filter for:
#    objectmethod foo (Int $i) {}
# where the parameter list is optional
$plugins->add_filter(
    Perl::Tidy::Sweetened::Keyword::Block->new(
        keyword     => 'objectmethod',
        marker      => 'OBJECTMETHOD',
        replacement => 'sub',
        clauses     => ['PAREN?'],
    ) );

# Create a subroutine filter for:
#    class Foo extends Bar {
#    class Foo with Bar, Baz {
# where both the extends and with are optional
$plugins->add_filter(
    Perl::Tidy::Sweetened::Keyword::Block->new(
        keyword     => 'class',
        marker      => 'CLASS',
        replacement => 'package',
        clauses =>
          [ '(with(\s+\w+)*)?', '(extends \s+ [\w|:]+)?', '(is(\s+\w+)*)?', ],
    ) );

# Create a twigil filter for:
#    $!variable_name
$plugins->add_filter(
    Perl::Tidy::Sweetened::Variable::Twigils->new(
        twigil => '$!',
        marker => 'TWG_BANG',
    ) );

sub perltidy {
    my %args = @_;
    $plugins->add_args( $args{argv} );

    return Perl::Tidy::perltidy(
        prefilter  => sub { $plugins->prefilter( $_[0] ) },
        postfilter => sub { $plugins->postfilter( $_[0] ) },
        @_
    );
}

1;

__END__

=pod

=head1 NAME

Perl::Tidy::Sweetened - Tweaks to Perl::Tidy to support some syntactic sugar

=head1 VERSION

version 1.19

=head1 STATUS

=for html <a href="https://travis-ci.org/mvgrimes/Perl-Tidy-Sweetened"><img src="https://travis-ci.org/mvgrimes/Perl-Tidy-Sweetened.svg?branch=master" alt="Build Status"></a>
<a href="https://metacpan.org/pod/Perl::Tidy::Sweetened"><img alt="CPAN version" src="https://badge.fury.io/pl/Perl-Tidy-Sweetened.svg" /></a>

=head1 DESCRIPTION

There are a number of modules on CPAN that allow users to write their classes
with a more "modern" syntax. These tools eliminate the need to shift off
C<$self>, can support type checking and offer other improvements.
Unfortunately, they can break the support tools that the Perl community has
come to rely on. This module attempts to work around those issues.

The module uses
L<Perl::Tidy>'s C<prefilter> and C<postfilter> hooks to support C<method> and
C<func> keywords, including the (possibly multi-line) parameter lists. This is
quite an ugly hack, but it is the recommended method of supporting these new
keywords (see the 2010-12-17 entry in the Perl::Tidy
L<CHANGES|https://metacpan.org/source/SHANCOCK/Perl-Tidy-20120714/CHANGES>
file). B<The resulting formatted code will leave the parameter lists untouched.>

C<Perl::Tidy::Sweetened> attempts to support the syntax outlined in the
following modules, but most of the new syntax styles should work:

=over

=item * p5-mop

=item * Method::Signatures::Simple

=item * MooseX::Method::Signatures

=item * MooseX::Declare

=item * Moops

=item * perl 5.20 signatures

=item * Kavorka

=back

=head1 THANKS

The idea and much of original code taken from Jonathan Swartz'
L<blog|http://www.openswartz.com/2010/12/19/perltidy-and-method-happy-together/>.

=head1 SEE ALSO

L<Perl::Tidy>

=head1 AUTHOR

Mark Grimes <mgrimes@cpan.org>

=head1 SOURCE

Source repository is at L<https://github.com/mvgrimes/Perl-Tidy-Sweetened>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://github.com/mvgrimes/Perl-Tidy-Sweetened/issues>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2023 by Mark Grimes <mgrimes@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
